core/modules/replica/IndexJob.cc

```diff
@@ -0,0 +1,538 @@
+    // locally created objects.
+
+    _chunks.clear();
+
+    // To ensure no lingering "side effects" will be left after cancelling this
+    // job the request cancellation should be also followed (where it makes a sense)
+    // by stopping the request at corresponding worker service.
+
+    for (auto&& itr: _requests) {
+        auto&& ptr = itr.second;
+        ptr->cancel();
+        if (ptr->state() != Request::State::FINISHED) {
+            controller()->stopById<StopIndexRequest>(
+                ptr->worker(),
+                ptr->id(),
+                nullptr,    /* onFinish */
+                options(lock).priority,
+                true,       /* keepTracking */
+                id()        /* jobId */
+            );
+        }
+    }
+    _requests.clear();
+    _rollbackTransaction(__func__);
+}
+
+
+void IndexJob::notify(util::Lock const& lock) {
+    LOGS(_log, LOG_LVL_DEBUG, context() << __func__);
+    notifyDefaultImpl<IndexJob>(lock, _onFinish);
+}
+
+
+void IndexJob::_onRequestFinish(IndexRequest::Ptr const& request) {
+
+    // NOTE: this algorithm assumes "zero tolerance" to failures - any failure
+    // in executing requests or processing data of the requests would result in
+    // the job termination. The only exception from this rule is a scenario
+    // when a target chunk table won't have a partition. This may be expected
+    // for some chunk tables because they may not have contributions in a context
+    // of the given super-transaction.
+    //
+    // TODO: reconsider this algorithm.
+
+    LOGS(_log, LOG_LVL_DEBUG, context() << __func__ << "  worker=" << request->worker());
+
+    if (state() == State::FINISHED) return;
+
+    util::Lock lock(_mtx, context() + __func__);
+
+    if (state() == State::FINISHED) return;
+
+    bool hasData = true;
+    if (request->extendedState() != Request::SUCCESS) {
+        if (request->extendedServerStatus() == EXT_STATUS_NO_SUCH_PARTITION) {
+            // OK to proceed. We just don't have any contribution into the
+            // partition.
+            hasData = false;
+        } else {
+            _resultData.error[request->worker()][request->chunk()] = request->responseData().error;
+            _rollbackTransaction(__func__);
+            finish(lock, ExtendedState::FAILED);
+            return;
+        }
+    }
+
+    // Submit a replacement request for the same worker BEFORE processing
+    // results of the current one. This little optimization is meant to keep
+    // workers busy in case of a non-negligible latency in processing data of
+    // requests.
+
+    for (auto&& ptr: _launchRequests(lock, request->worker())) {
+        _requests[ptr->id()] = ptr;
+    }
+
+    // Removing request from the list before processing its data is fine as
+    // we still have a shared pointer passed into this method. Note that
+    // we need to erase completed requests from memory since they may carry
+    // a significant amount of data.
```

Comment:
Its probably safer (and it doesn't cost that much) for this function to have it's own copy of `request` if its going to erase the main pointer.